src/main/java/val/ResultMap.java

```diff
@@ -0,0 +1,167 @@
+package val;
+
+import java.util.*;
+
+import org.checkerframework.checker.nullness.qual.*;
+
+/**
+ * An immutable value object which conveys the result of an evaluation
+ * of a {@link Check} against a provided input.
+ * This is represented as a Map where the `key` defines the context
+ * for specific feedback and the value is a {@link Result} object
+ * providing that feedback. In conventional use the key will be the
+ * field or complete path to a field for which the {@link Result}
+ * applies.
+ *
+ * If the map is empty then the checks have nothing to say (everything passed)
+ * and the ResultMap is considered `clean`.
+ */
+public final class ResultMap {
+    /**
+     * The results this ResultMap represents.
+     *
+     * The keys identify the path or similar context
+     * and the values are the list of {@link Result}s for that context.
+     */
+    private final Map<String, List<Result>> values;
+
+    private final boolean clean;
```

Comment:
why bother storing this rather than just returning `this.values.isEmpty()`?

Reply:
good question...I have no idea. I mostly copied this code from something I wrote a while ago so whatever reason I had is lost in the sands of time (likely either some over optimized attempt to avoid that extra indirection, some mistrust of values, or some envisioned future extension). I'll adjust it.

Reply:
I think the reason (or at least a reason) for this came back to me. This was designed as an optimization, largely to avoid the comparison that Map#isEmpty normally does internally. By itself this is inconsequential, but this method is used extensively throughout the framework (in pretty much all of the Check composition functions) and most of the time it will be done against CLEAN so this would avoid all of those extra comparisons which are unlikely to be too automatically optimized due to the underlying Map object. This should be better documented and I'm also not a fan of this particular solution for that problem. I'll push an alternative where that method logic is more directly overridden in the CLEAN instance.

Reply:
Replaced with slicker approach and better comments